tests/numerics/ad/test_secondary_operator.py

```diff
@@ -0,0 +1,300 @@
+                    return eqsys.md_variable(name, subdomains)
+                elif all(isinstance(d, pp.BoundaryGrid) for d in subdomains):
+                    return pp.ad.TimeDependentDenseArray(name, subdomains)
+
+        elif name in [INTFVAR_NAME]:
+
+            def _var(interfaces):
+                return eqsys.md_variable(name, interfaces)
+
+        else:
+            raise NotImplementedError(f"get_var fixture not supporting var name {name}")
+
+        return _var
+
+    return _get_var
+
+
+@pytest.mark.parametrize("on_intf", [False, True])
+def test_secondary_operators(
+    on_intf: bool,
+    eqsys: pp.ad.EquationSystem,
+    get_var: Callable[[str], Callable[[pp.GridLikeSequence], pp.ad.Operator]],
+):
+    """Test all aspects of the secondary operator at current time and iter."""
+
+    mdg = eqsys.mdg
+    sds = mdg.subdomains()
+    intfs = mdg.interfaces()
+    bgs = mdg.boundaries()
+
+    if on_intf:
+        nc = mdg.num_interface_cells()
+        domains = intfs
+        vars = [get_var(INTFVAR_NAME)(domains)]
+        diff_vals = [2 * np.ones(nc)]
+        expr = pp.ad.SecondaryExpression(
+            "interface_expression",
+            eqsys.mdg,
+            [get_var(INTFVAR_NAME)],
+            time_dependent=True,
+        )
+    else:
+        nc = mdg.num_subdomain_cells()
+        domains = sds
+        vars = [get_var(VAR1_NAME)(domains), get_var(VAR2_NAME)(domains)]
+
+        diff_vals = [2 * np.ones(nc), 3 * np.ones(nc)]
+        expr = pp.ad.SecondaryExpression(
+            "subdomain_expression",
+            eqsys.mdg,
+            [get_var(VAR1_NAME), get_var(VAR2_NAME)],
+            time_dependent=True,
+        )
+
+    # Testing the created operators
+    sop = expr(domains)
+    sop_empty = expr([])
+
+    sop_pi = sop.previous_iteration()
+    sop_pt = sop.previous_timestep()
+
+    # Test that the right operator was created
+    assert isinstance(sop, pp.ad.SecondaryOperator)
+    # Calling the secondary expression with no grids, gives a wrapped empty array
+    assert isinstance(sop_empty, pp.ad.DenseArray)
+    assert sop_empty.value(eqsys).shape == (0,)
+
+    # secondary operator (SOP) starts at current iterate
+    assert sop.time_step_index == 0
+    assert sop.iterate_index == 0
+    # assert the prev index was increased for SOP and its children
+    assert sop_pi.iterate_index == 1
+    assert sop_pi.time_step_index == 0
+    assert all(v.iterate_index == 1 for v in sop_pi.children)
+    assert all(v.time_step_index == 0 for v in sop_pi.children)
+    assert sop_pt.iterate_index == 0
+    assert sop_pt.time_step_index == 1
+    assert all(v.iterate_index == 0 for v in sop_pt.children)
+    assert all(v.time_step_index == 1 for v in sop_pt.children)
+
+    # check that arithmetic overloads work for secondary operator, even though it
+    # inherits from AbstractFunction (correct MOR)
+    a = sop + vars[0]
+    try:
+        for a in ["+", "-", "*", "@", "**"]:
```

Comment:
If the `SecondaryOperator` is a genuinely new type of `Operator` (and we need to discuss this), I would prefer that correct parsing of arithmetic operations are included in the [general test for such methods](https://github.com/pmgbergen/porepy/blob/0e94301536d6640cea9ae91816d368d8efa90d75/tests/numerics/ad/test_operators.py#L1896). This would require a careful definition of which combinations of arithmetic operations and operands can be permitted for a `SecondaryOperator`, but that will be required for any new `Operator.`

Reply:
Discussed in person, removing this part.